File: tugasHeap.py
import csv
 
min_heap = []
max_heap = []
 
def min_naik(heap, i):
    while i > 0:
        parent = (i - 1) // 2
        if heap[i][0] < heap[parent][0]:
            heap[i], heap[parent] = heap[parent], heap[i]
            i = parent
        else:
            break
 
def min_turun(heap, i):
    n = len(heap)
    while True:
        terkecil = i
        kiri, kanan = 2*i+1, 2*i+2
        if kiri  < n and heap[kiri][0]  < heap[terkecil][0]: terkecil = kiri
        if kanan < n and heap[kanan][0] < heap[terkecil][0]: terkecil = kanan
        if terkecil == i: break
        heap[i], heap[terkecil] = heap[terkecil], heap[i]
        i = terkecil
 
def tambah_min(id, nama):
    min_heap.append([id, nama])
    min_naik(min_heap, len(min_heap) - 1)
 
def hapus_min(id):
    for i in range(len(min_heap)):
        if min_heap[i][0] == id:
            nama = min_heap[i][1]
            min_heap[i] = min_heap[-1]
            min_heap.pop()
            if i < len(min_heap):
                min_naik(min_heap, i)
                min_turun(min_heap, i)
            print(f"✔ Dihapus dari MinHeap: [{id}, {nama}]")
            return
    print(f"✖ ID {id} tidak ditemukan di MinHeap.")
 
def max_naik(heap, i):
    while i > 0:
        parent = (i - 1) // 2
        if heap[i][0] > heap[parent][0]:
            heap[i], heap[parent] = heap[parent], heap[i]
            i = parent
        else:
            break
 
def max_turun(heap, i):
    n = len(heap)
    while True:
        terbesar = i
        kiri, kanan = 2*i+1, 2*i+2
        if kiri  < n and heap[kiri][0]  > heap[terbesar][0]: terbesar = kiri
        if kanan < n and heap[kanan][0] > heap[terbesar][0]: terbesar = kanan
        if terbesar == i: break
        heap[i], heap[terbesar] = heap[terbesar], heap[i]
        i = terbesar
 
def tambah_max(id, nama):
    max_heap.append([id, nama])
    max_naik(max_heap, len(max_heap) - 1)
 
def hapus_max(id):
    for i in range(len(max_heap)):
        if max_heap[i][0] == id:
            nama = max_heap[i][1]
            max_heap[i] = max_heap[-1]
            max_heap.pop()
            if i < len(max_heap):
                max_naik(max_heap, i)
                max_turun(max_heap, i)
            print(f"✔ Dihapus dari MaxHeap: [{id}, {nama}]")
            return
    print(f"✖ ID {id} tidak ditemukan di MaxHeap.")
 
def tampil_ascending():
    if not min_heap:
        print("Heap kosong!"); return
    tmp = [x[:] for x in min_heap]
    print(f"\n{'No':<5} {'ID':<8} Nama")
    print("-" * 30)
    no = 1
    while tmp:
        idx = min(range(len(tmp)), key=lambda i: tmp[i][0])
        item = tmp.pop(idx)
        print(f"{no:<5} {item[0]:<8} {item[1]}")
        no += 1
 
def tampil_descending():
    if not max_heap:
        print("Heap kosong!"); return
    tmp = [x[:] for x in max_heap]
    print(f"\n{'No':<5} {'ID':<8} Nama")
    print("-" * 30)
    no = 1
    while tmp:
        idx = max(range(len(tmp)), key=lambda i: tmp[i][0])
        item = tmp.pop(idx)
        print(f"{no:<5} {item[0]:<8} {item[1]}")
        no += 1
 
def load_csv(filename):
    try:
        with open(filename, newline='', encoding='utf-8') as file:
            reader = csv.reader(file)
            next(reader)  # skip header
 
            for row in reader:
                if len(row) < 2:
                    continue
 
                # support ; atau ,
                if ";" in row[0]:
                    row = row[0].split(";")
 
                id   = int(row[0].strip())
                nama = row[1].strip()
 
                tambah_min(id, nama)
                tambah_max(id, nama)
 
        print(f"✅ Semua data berhasil dimasukkan ke Min-Heap & Max-Heap!")
 
    except Exception as e:
        print("❌ Error:", e)
 
while True:
    print("\n=== MENU HEAP (PYTHON) ===")
    print("1. Load Data dari CSV")
    print("2. Tambah Data")
    print("3. Tampil ASCENDING  (Min-Heap)")
    print("4. Tampil DESCENDING (Max-Heap)")
    print("5. Hapus dari Min-Heap")
    print("6. Hapus dari Max-Heap")
    print("0. Keluar")
 
    pilih = input("Pilih: ")
 
    if pilih == "1":
        load_csv("D:/kuliah/semester 4/Struktur Data/praktikum/#7 Heap/data100.csv")
 
    elif pilih == "2":
        id   = int(input("ID   : "))
        nama = input("Nama : ")
        tambah_min(id, nama)
        tambah_max(id, nama)
        print(f"✔ [{id}, {nama}] berhasil ditambahkan!")
 
    elif pilih == "3":
        tampil_ascending()
 
    elif pilih == "4":
        tampil_descending()
 
    elif pilih == "5":
        id = int(input("Masukkan ID: "))
        hapus_min(id)
 
    elif pilih == "6":
        id = int(input("Masukkan ID: "))
        hapus_max(id)
 
    elif pilih == "0":
        print("Program selesai.")
        break
 
    else:
        print("Pilihan tidak valid!")

File: tugasHeap.java
import java.util.*;
import java.io.*;
 
public class heap {
 
    static int[]    minId   = new int[1000];
    static String[] minNama = new String[1000];
    static int      minSize = 0;
 
    static int[]    maxId   = new int[1000];
    static String[] maxNama = new String[1000];
    static int      maxSize = 0;
 
    static void tambahMin(int id, String nama) {
        minId[minSize]   = id;
        minNama[minSize] = nama;
        int i = minSize;
        while (i > 0 && minId[i] < minId[(i-1)/2]) {
            int t = minId[i]; minId[i] = minId[(i-1)/2]; minId[(i-1)/2] = t;
            String s = minNama[i]; minNama[i] = minNama[(i-1)/2]; minNama[(i-1)/2] = s;
            i = (i-1)/2;
        }
        minSize++;
    }
 
    static void hapusMin(int id) {
        int idx = -1;
        for (int i = 0; i < minSize; i++)
            if (minId[i] == id) { idx = i; break; }
 
        if (idx == -1) { System.out.println("✖ ID " + id + " tidak ditemukan di MinHeap."); return; }
 
        System.out.println("✔ Dihapus dari MinHeap: [" + id + ", " + minNama[idx] + "]");
        minSize--;
        minId[idx] = minId[minSize]; minNama[idx] = minNama[minSize];
        int i = idx;
        while (true) {
            int terkecil = i, kiri = 2*i+1, kanan = 2*i+2;
            if (kiri  < minSize && minId[kiri]  < minId[terkecil]) terkecil = kiri;
            if (kanan < minSize && minId[kanan] < minId[terkecil]) terkecil = kanan;
            if (terkecil == i) break;
            int t = minId[i]; minId[i] = minId[terkecil]; minId[terkecil] = t;
            String s = minNama[i]; minNama[i] = minNama[terkecil]; minNama[terkecil] = s;
            i = terkecil;
        }
    }
 
    static void tambahMax(int id, String nama) {
        maxId[maxSize]   = id;
        maxNama[maxSize] = nama;
        int i = maxSize;
        while (i > 0 && maxId[i] > maxId[(i-1)/2]) {
            int t = maxId[i]; maxId[i] = maxId[(i-1)/2]; maxId[(i-1)/2] = t;
            String s = maxNama[i]; maxNama[i] = maxNama[(i-1)/2]; maxNama[(i-1)/2] = s;
            i = (i-1)/2;
        }
        maxSize++;
    }
 
    static void hapusMax(int id) {
        int idx = -1;
        for (int i = 0; i < maxSize; i++)
            if (maxId[i] == id) { idx = i; break; }
 
        if (idx == -1) { System.out.println("✖ ID " + id + " tidak ditemukan di MaxHeap."); return; }
 
        System.out.println("✔ Dihapus dari MaxHeap: [" + id + ", " + maxNama[idx] + "]");
        maxSize--;
        maxId[idx] = maxId[maxSize]; maxNama[idx] = maxNama[maxSize];
        int i = idx;
        while (true) {
            int terbesar = i, kiri = 2*i+1, kanan = 2*i+2;
            if (kiri  < maxSize && maxId[kiri]  > maxId[terbesar]) terbesar = kiri;
            if (kanan < maxSize && maxId[kanan] > maxId[terbesar]) terbesar = kanan;
            if (terbesar == i) break;
            int t = maxId[i]; maxId[i] = maxId[terbesar]; maxId[terbesar] = t;
            String s = maxNama[i]; maxNama[i] = maxNama[terbesar]; maxNama[terbesar] = s;
            i = terbesar;
        }
    }
 
    static void tampilAscending() {
        if (minSize == 0) { System.out.println("Heap kosong!"); return; }
        int[]    tmpId   = Arrays.copyOf(minId,   minSize);
        String[] tmpNama = Arrays.copyOf(minNama, minSize);
        int      tmpSize = minSize;
 
        System.out.printf("%n%-5s %-8s %s%n", "No", "ID", "Nama");
        System.out.println("-".repeat(30));
        for (int no = 1; no <= minSize; no++) {
            int idx = 0;
            for (int i = 1; i < tmpSize; i++)
                if (tmpId[i] < tmpId[idx]) idx = i;
            System.out.printf("%-5d %-8d %s%n", no, tmpId[idx], tmpNama[idx]);
            tmpId[idx] = tmpId[tmpSize-1]; tmpNama[idx] = tmpNama[tmpSize-1]; tmpSize--;
        }
    }
 
    static void tampilDescending() {
        if (maxSize == 0) { System.out.println("Heap kosong!"); return; }
        int[]    tmpId   = Arrays.copyOf(maxId,   maxSize);
        String[] tmpNama = Arrays.copyOf(maxNama, maxSize);
        int      tmpSize = maxSize;
 
        System.out.printf("%n%-5s %-8s %s%n", "No", "ID", "Nama");
        System.out.println("-".repeat(30));
        for (int no = 1; no <= maxSize; no++) {
            int idx = 0;
            for (int i = 1; i < tmpSize; i++)
                if (tmpId[i] > tmpId[idx]) idx = i;
            System.out.printf("%-5d %-8d %s%n", no, tmpId[idx], tmpNama[idx]);
            tmpId[idx] = tmpId[tmpSize-1]; tmpNama[idx] = tmpNama[tmpSize-1]; tmpSize--;
        }
    }
 
    static void loadCSV(String fileName) {
        try (BufferedReader br = new BufferedReader(new FileReader(fileName))) {
            String line;
            br.readLine(); // skip header
 
            while ((line = br.readLine()) != null) {
                String[] data;
 
                // support ; atau ,
                if (line.contains(";"))
                    data = line.split(";");
                else
                    data = line.split(",");
 
                int id      = Integer.parseInt(data[0].trim());
                String nama = data[1].trim();
 
                tambahMin(id, nama);
                tambahMax(id, nama);
            }
 
            System.out.println("✔ Semua data berhasil dimasukkan ke Min-Heap & Max-Heap!");
 
        } catch (Exception e) {
            System.out.println("❌ Error: " + e.getMessage());
        }
    }
 
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int pilih;
 
        do {
            System.out.println("\n=== MENU HEAP (JAVA) ===");
            System.out.println("1. Load Data dari CSV");
            System.out.println("2. Tambah Data");
            System.out.println("3. Tampil ASCENDING  (Min-Heap)");
            System.out.println("4. Tampil DESCENDING (Max-Heap)");
            System.out.println("5. Hapus dari Min-Heap");
            System.out.println("6. Hapus dari Max-Heap");
            System.out.println("0. Keluar");
            System.out.print("Pilih: ");
            pilih = sc.nextInt();
 
            switch (pilih) {
                case 1:
                    loadCSV("D:/kuliah/semester 4/Struktur Data/praktikum/#7 Heap/data100.csv");
                    break;
 
                case 2:
                    System.out.print("ID   : ");
                    int id = sc.nextInt(); sc.nextLine();
                    System.out.print("Nama : ");
                    String nama = sc.nextLine();
                    tambahMin(id, nama);
                    tambahMax(id, nama);
                    System.out.println("✔ [" + id + ", " + nama + "] berhasil ditambahkan!");
                    break;
 
                case 3:
                    tampilAscending();
                    break;
 
                case 4:
                    tampilDescending();
                    break;
 
                case 5:
                    System.out.print("Masukkan ID: ");
                    hapusMin(sc.nextInt());
                    break;
 
                case 6:
                    System.out.print("Masukkan ID: ");
                    hapusMax(sc.nextInt());
                    break;
 
                case 0:
                    System.out.println("Program selesai.");
                    break;
 
                default:
                    System.out.println("Pilihan tidak valid!");
            }
 
        } while (pilih != 0);
 
        sc.close();
    }
}
